spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -0,0 +1,180 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.core;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.Entity.Builder;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.Query;
+import com.google.common.annotations.VisibleForTesting;
+
+import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreEntityConverter;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntity;
+import org.springframework.data.mapping.PersistentProperty;
+import org.springframework.util.Assert;
+
+/**
+ * An implementation of {@link DatastoreOperations}.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class DatastoreTemplate implements DatastoreOperations {
+
+	private final Datastore datastore;
+
+	private final DatastoreEntityConverter datastoreEntityConverter;
+
+	private final DatastoreMappingContext datastoreMappingContext;
+
+	public DatastoreTemplate(Datastore datastore,
+			DatastoreEntityConverter datastoreEntityConverter,
+			DatastoreMappingContext datastoreMappingContext) {
+		Assert.notNull(datastore, "A non-null Datastore service object is required.");
+		Assert.notNull(datastoreEntityConverter,
+				"A non-null DatastoreEntityConverter is required.");
+		Assert.notNull(datastoreMappingContext,
+				"A non-null DatastoreMappingContext is required.");
+		this.datastore = datastore;
+		this.datastoreEntityConverter = datastoreEntityConverter;
+		this.datastoreMappingContext = datastoreMappingContext;
+	}
+
+	@VisibleForTesting
+	Key getKeyFromId(Object id, Class entityClass) {
+		Assert.notNull(id, "Cannot get key for null ID value.");
+		if (id instanceof Key) {
+			return (Key) id;
+		}
+		KeyFactory keyFactory = this.datastore.newKeyFactory();
+		keyFactory.setKind(
+				this.datastoreMappingContext.getPersistentEntity(entityClass).kindName());
+		Key key;
+		if (id instanceof String) {
+			key = keyFactory.newKey((String) id);
+		}
+		else if (id == long.class) {
```

Comment:
Does this work? Why not `id instanceof Long`
Also, what about `Integer`?

Reply:
Oh my goodness, my bad, will fix.

Regarding not including Integer or int or Long: these will be handled by the custom converters and the Spring Data conversion service that we will introduce later. So, I don't want to include these handful of similar-ish types because that will be redundant because the conversion service we will be using by default handles 90 or so of these common type conversions.